### PR TITLE
[CANNOLI-29] Add minimal GMAP and GSNAP wrappers.

### LINKED ISSUE #29: Add wrappers for GMAP and GSNAP aligners
http://research-pub.gene.com/gmap/

### PR DESCRIPTION
**Pull request by heuermh**
Fixes #29.

### CODE REVIEW

File: src/main/scala/org/bdgenomics/cannoli/Gmap.scala [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  val companion = Gmap
+  val stringency = ValidationStringency.valueOf(args.stringency)
+
+  def run(sc: SparkContext) {
+    val input: FragmentRDD = sc.loadFragments(args.inputPath)
+
+    implicit val tFormatter = InterleavedFASTQInFormatter
+    implicit val uFormatter = new AnySAMOutFormatter
+
+    val gmapCommand = "gmap --dir= " + args.genomePath + " --db=" + args.genomeName + " --format=sampe"
```
Comment (fnothaft) [inline src/main/scala/org/bdgenomics/cannoli/Gmap.scala]: I'd make the path to `gmap` parametrizable.

File: src/main/scala/org/bdgenomics/cannoli/Gsnap.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  val companion = Gsnap
+  val stringency = ValidationStringency.valueOf(args.stringency)
+
+  def run(sc: SparkContext) {
+    val input: FragmentRDD = sc.loadFragments(args.inputPath)
+
+    implicit val tFormatter = InterleavedFASTQInFormatter
+    implicit val uFormatter = new AnySAMOutFormatter
+
+    val gsnapCommand = "gsnap --dir= " + args.genomePath + " --db=" + args.genomeName + " --format=sam"
```
Comment (fnothaft) [inline src/main/scala/org/bdgenomics/cannoli/Gsnap.scala]: I'd make the path to `gsnap` parametrizable.

**Comment by heuermh** [comment]
> Let's chat tomorrow about getting "feature complete" across all the aligners (e.g., docker support, etc).

+1

File: src/main/scala/org/bdgenomics/cannoli/Gmap.scala [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    implicit val tFormatter = InterleavedFASTQInFormatter
+    implicit val uFormatter = new AnySAMOutFormatter
+
+    val gmapCommand = Seq(args.gmapPath,
+      "--dir=" + args.genomePath,
+      "--db=" + args.genomeName,
+      "--format=sampe").mkString(" ")
+
+    val output: AlignmentRecordRDD = input.pipe[AlignmentRecord, AlignmentRecordRDD, InterleavedFASTQInFormatter](gmapCommand)
+      .transform(_.cache())
```
Comment (fnothaft) [inline src/main/scala/org/bdgenomics/cannoli/Gmap.scala]: Why are we caching here, if we're saving direct to disk?

File: src/main/scala/org/bdgenomics/cannoli/Gsnap.scala [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    implicit val tFormatter = InterleavedFASTQInFormatter
+    implicit val uFormatter = new AnySAMOutFormatter
+
+    val gsnapCommand = Seq(args.gsnapPath,
+      "--dir=" + args.genomePath,
+      "--db=" + args.genomeName,
+      "--format=sam").mkString(" ")
+
+    val output: AlignmentRecordRDD = input.pipe[AlignmentRecord, AlignmentRecordRDD, InterleavedFASTQInFormatter](gsnapCommand)
+      .transform(_.cache())
```
Comment (fnothaft) [inline src/main/scala/org/bdgenomics/cannoli/Gsnap.scala]: Ditto here. Why the `cache`?

**Comment by heuermh** [comment]
This one is ready to go.  Will need docker support and index mapping as described in #34.

**Comment by heuermh** [comment]
Rebased.

**Comment by heuermh** [comment]
> Do these two tools accept interleaved FASTQ on stdin by default?

I should've read closer, they accept FASTQ on stdin by default but assume it to be single-end reads
```
Input to GSNAP should be either in FASTQ or FASTA format.  The FASTQ
input may include quality scores, which will then be included in SAM
output, if that output format is selected.  For single-end reads, the
FASTQ file may be piped into GSNAP, or given as its command-line
argument, like this

    cat <fastq_file> | gsnap -d <genome>

or

    gsnap -d <genome> <fastq_file>


For paired-end reads, the two corresponding FASTQ files should be
given as command-line arguments in pairs, like this

    gsnap -d <genome> <fastq_file_1> <fastq_file_2> [<fastq_file_3> <fastq_file_4>...]

A pipe cannot work since GSNAP needs to access both FASTQ files in
parallel.  The reads in FASTQ files may have varying lengths, if
desired.  Note that GSNAP can process multiple sets of paired-end
reads, by adding the files in pairs.  If you want to provide multiple
single-end files, you can either use "cat" to concatenate them into
the stdin of gsnap, like this:

    cat <fastq_file_1> [<fastq_file_2>...] | gsnap -d <genome>

or you can provide them all on the command line with the
--force-single-end flag, like this:

    gsnap -d <genome> --force-single-end <fastq_file_1> [<fastq_file_2>...]

which will process each FASTQ file one at a time as single-end reads,
and not try to pair them up.
```

**Comment by fnothaft** [comment]
@heuermh thoughts on a path forward?

**Comment by heuermh** [comment]
Not having any issue tracker, I've sent an email to the author.  Unless you can think of another approach, like splitting the interleaved reads into separate named pipes in a wrapper script.

**Comment by heuermh** [comment]
Closing as WontFix, no reply from author.

**Pull request closed without merging**